File: admin/src/utils/RsaUtils.js
/*
 * [RsaUtils.js]
 * --------------------------------------------------------------------------------
 * This software is licensed under the MIT License.
 * However, any distribution or modification must retain this copyright notice.
 * See LICENSE for full terms.
 * --------------------------------------------------------------------------------
 * author: "Jiu Liu"
 * author_contact: "QQ: 3209174373, GitHub: https://github.com/DCSCDF"
 * license: "MIT"
 * license_exception: "Mandatory attribution retention"
 * UpdateTime: 2026/2/18 00:43
 *
 */

import JSEncrypt from 'jsencrypt';

/**
 * RSA 加密工具类（兼容 Java 后端 PKCS1Padding）
 */
class RsaEncryptor {
	/**
	 * 使用 Base64 编码的公钥对明文进行 RSA 加密
	 * @param {string} plaintext - 要加密的明文（如密码）
	 * @param {string} publicKeyBase64 - 后端提供的 Base64 编码的公钥（PKCS#8 格式？注意：JSEncrypt 需要 PKCS#1 格式的公钥）
	 * @returns {string} Base64 编码的密文
	 */
	static encrypt(plaintext, publicKeyBase64) {
		if (!plaintext || !publicKeyBase64) {
			throw new Error('明文或公钥不能为空');
		}

		//Java 生成的公钥是 X.509 SubjectPublicKeyInfo 格式（PKCS#8 公钥结构）
		// JSEncrypt 只接受 PKCS#1 格式的公钥（即以 -----BEGIN RSA PUBLIC KEY----- 开头）
		// 但幸运的是，JSEncrypt 内部可以自动处理 X.509 格式的公钥（以 -----BEGIN PUBLIC KEY----- 开头）
		// 所以我们只需将 Base64 转为 PEM 格式即可


		const pemPublicKey = RsaEncryptor.base64ToPem(publicKeyBase64, 'PUBLIC KEY');

		const encryptor = new JSEncrypt();
		encryptor.setPublicKey(pemPublicKey);

		const encrypted = encryptor.encrypt(plaintext);

		if (!encrypted) {
			throw new Error('RSA 加密失败，请检查公钥格式或明文长度');
		}

		return encrypted; // JSEncrypt.encrypt() 返回的就是 Base64 字符串
	}

	/**
	 * 将 Base64 编码的密钥转换为 PEM 格式
	 * @param {string} base64Key
	 * @param {string} keyType - 'PUBLIC KEY' 或 'RSA PRIVATE KEY' 等
	 * @returns {string} PEM 格式的密钥
	 */
	static base64ToPem(base64Key, keyType) {
		const lines = [];
		const key = base64Key.replace(/\s/g, ''); // 去除空格
		const prefix = `-----BEGIN ${keyType}-----\n`;
		const suffix = `\n-----END ${keyType}-----`;

		// 每 64 个字符一行（PEM 标准）
		for (let i = 0; i < key.length; i += 64) {
			lines.push(key.substring(i, i + 64));
		}

		return prefix + lines.join('\n') + suffix;
	}
}

export default RsaEncryptor;

File: server/src/main/java/com/jiuliu/myblog_dev/utils/rsa/RsaUtils.java
/*
 * [RsaUtils.java]
 * =======================================
 * This software is licensed under the MIT License.
 * However, any distribution or modification must retain this copyright notice.
 * See LICENSE for full terms.
 * =======================================
 * author: "Jiu Liu"
 * author_contact: "QQ: 3209174373, GitHub: https://github.com/DCSCDF"
 * license: "MIT"
 * license_exception: "Mandatory attribution retention"
 * UpdateTime: 2026/3/8 04:37
 */

package com.jiuliu.myblog_dev.utils.rsa;


import com.jiuliu.myblog_dev.exception.DecryptionException;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.crypto.Cipher;
import java.nio.charset.StandardCharsets;
import java.security.KeyFactory;
import java.security.PrivateKey;
import java.security.spec.PKCS8EncodedKeySpec;
import java.util.Base64;

public class RsaUtils {

    private static final Logger log = LoggerFactory.getLogger(RsaUtils.class);

    /**
     * 使用私钥解密 Base64 编码的 RSA 密文
     *
     * @param encryptedBase64  Base64 编码的密文（来自前端）
     * @param privateKeyBase64 Base64 编码的 PKCS#8 私钥
     * @return 解密后的明文
     * @throws DecryptionException 当解密失败时（如数据损坏、非密文、密钥不匹配等）
     */
    public static String decryptByPrivateKey(String encryptedBase64, String privateKeyBase64) {
        if (encryptedBase64 == null || encryptedBase64.isEmpty()) {
            log.warn("RSA 解密失败：输入密文为空");
            throw new DecryptionException("无效的加密数据");
        }

        try {
            // 解码私钥
            byte[] privateKeyBytes = Base64.getDecoder().decode(privateKeyBase64);
            PKCS8EncodedKeySpec keySpec = new PKCS8EncodedKeySpec(privateKeyBytes);
            KeyFactory keyFactory = KeyFactory.getInstance("RSA");
            PrivateKey privateKey = keyFactory.generatePrivate(keySpec);

            // 初始化解密器
            Cipher cipher = Cipher.getInstance("RSA/ECB/PKCS1Padding");
            cipher.init(Cipher.DECRYPT_MODE, privateKey);

            // 解码并解密
            byte[] encryptedBytes = Base64.getDecoder().decode(encryptedBase64);
            byte[] decryptedBytes = cipher.doFinal(encryptedBytes);

            String result = new String(decryptedBytes, StandardCharsets.UTF_8);
            log.debug("RSA 解密成功，明文长度: {}", result.length());
            return result;

        } catch (IllegalArgumentException e) {
            // Base64 解码失败
            log.warn("RSA 解密失败：输入数据不是有效的 Base64");
            throw new DecryptionException("加密数据格式错误");
        } catch (Exception e) {
            // 包括 BadPaddingException, IllegalBlockSizeException, InvalidKeyException 等
            log.warn("RSA 解密失败：{}", e.getMessage());
            // 不记录完整堆栈，避免日志膨胀和潜在信息泄露
            throw new DecryptionException("解密失败，请刷新页面重试");
        }
    }
}